### ISSUE TITLE
torch.cat bug/unexpected behavior

### ISSUE THREAD

**Issue by thuyen** [pytorch/pytorch#627]
I would expect `x` and `y` are the same in the following:

````python
import numpy as np
import torch

a0 = np.reshape(np.arange(24), (4,6))
a1 = np.transpose(a0)

t1 = torch.from_numpy(a1)
x = t1.unsqueeze(0)[0]
y = torch.cat([t1.unsqueeze(0)])[0]

print(x)
print(y)
````

But they're not:
````
x
  0   6  12  18
  1   7  13  19
  2   8  14  20
  3   9  15  21
  4  10  16  22
  5  11  17  23
[torch.LongTensor of size 6x4]

y
  0   1   2   3
  4   5   6   7
  8   9  10  11
 12  13  14  15
 16  17  18  19
 20  21  22  23
[torch.LongTensor of size 6x4]
````

**Comment by apaszke** [comment]
Ugh. There's even a simpler repro:
```python
x = torch.range(1, 25).view(5, 5)
torch.cat([x.t()])
```
Weird thing is that it only breaks for `dim=0` (the default). With `dim=1` it's good 😕

**Comment by apaszke** [comment]
And it only happens if there's only one concatenated tensor. It must be a bug somewhere in the C libraries.

**Comment by thuyen** [comment]
Yes, right now the [dataloader](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L64) would behave incorrectly if the `batch_size=1` and we have any `transpose` during data transformation. At least this was how I encountered the problem.

**Comment by apaszke** [comment]
I see. Of course I'm not saying that it shouldn't be fixed, I'm just pointing out that it's quite a rare thing.

**Comment by thuyen** [comment]
Sure, just wanted to provide little context :)

**Comment by apaszke** [comment]
Fixed in #650.
